### PR TITLE
Filtros hotel y roomType

### PR DESCRIPTION
**Pull request by Javiruizg**
## User description
Filtros nuevos de hoteles y rooms. Se han hecho dos nuevos endpoints y tests que comprueban su funcionamiento

## PR Type
Enhancement, Tests

## Description
Added new filters (start_date, end_date) for hotel and room type queries.

Introduced two new endpoints for hotel and room type availability.

Implemented comprehensive tests for the new endpoints and filters.

Enhanced room type filtering logic with additional parameters.

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Introduced advanced filtering for hotels and room types.
  - Added new filters like `start_date`, `end_date`, and price range.
  - Refactored filter validation and application logic.

- Added two new endpoints for hotel and room type availability.
  - `available_hotels` endpoint to fetch hotels with available room types.
  - `available_room_types` endpoint to fetch available room types for a hotel.

- Implemented comprehensive tests for new endpoints and filtering logic.
  - Verified availability logic for hotels and room types.
  - Tested edge cases for overlapping and non-overlapping bookings.

- Updated permission services to include new endpoints.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>services.py</strong><dd><code>Refactor and enhance hotel filtering logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

backend/pawtel/hotels/services.py

<li>Refactored hotel filtering logic with validation and application <br>methods.<br> <li> Added support for new filters like <code>start_date</code>, <code>end_date</code>, and price <br>range.<br> <li> Introduced <code>list_filtered_hotels</code> method for advanced filtering.

</details>

  </td>
  <td><a href="https://github.com/LuisMelladoDiaz/Pawtel-ComparadorDeHotelesParaMascotas/pull/203/files#diff-095282d677bd0ecb4cfb20d05ac56a005b0510c195f4938da3adc3aa1e7b53f7">+91/-76</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>views.py</strong><dd><code>Add availability endpoints and update hotel listing</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

backend/pawtel/hotels/views.py

<li>Integrated new <code>available_hotels</code> and <code>available_room_types</code> endpoints.<br> <li> Updated hotel listing to use advanced filtering.

</details>

  </td>
  <td><a href="https://github.com/LuisMelladoDiaz/Pawtel-ComparadorDeHotelesParaMascotas/pull/203/files#diff-29eae86551ee9fe269942e1ca9a284e2c7c05aa7b4f183bd215252e670bce1fd">+49/-1</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>permission_services.py</strong><dd><code>Update permissions for new endpoints</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

backend/pawtel/permission_services.py

<li>Added permissions for <code>available_hotels</code> and <code>available_room_types</code> <br>endpoints.

</details>

  </td>
  <td><a href="https://github.com/LuisMelladoDiaz/Pawtel-ComparadorDeHotelesParaMascotas/pull/203/files#diff-4ed9e53a8c4f669655e92a60628026f2fc83490b4ca3d292338afd9ac43edcf5">+6/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>services.py</strong><dd><code>Add filtering logic for room types</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

backend/pawtel/room_types/services.py

<li>Added advanced filtering logic for room types.<br> <li> Introduced validation and application methods for filters.<br> <li> Implemented <code>list_filtered_room_types</code> method.

</details>

  </td>
  <td><a href="https://github.com/LuisMelladoDiaz/Pawtel-ComparadorDeHotelesParaMascotas/pull/203/files#diff-d37634bc6a11424d1ff1ac529d0c1bb1643125882678b1ab117d9b5a9def629a">+68/-1</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_views3.py</strong><dd><code>Add tests for availability endpoints</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

backend/pawtel/hotels/tests/test_views3.py

<li>Added tests for <code>available_hotels</code> and <code>available_room_types</code> endpoints.<br> <li> Verified hotel and room type availability logic.<br> <li> Tested edge cases for overlapping bookings.

</details>

  </td>
  <td><a href="https://github.com/LuisMelladoDiaz/Pawtel-ComparadorDeHotelesParaMascotas/pull/203/files#diff-5c7c36d8a4ad437b2bc2c78c843432add0d0b3195ab4f04d2ded535011a9f3fe">+152/-0</a>&nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about Qodo Merge usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

### CODE REVIEW

File: backend/pawtel/hotels/views.py [hunk 1]
```diff
@@ -98,6 +99,53 @@ def list_bookings_of_hotel(self, request, pk=None):
...
+                    break
+        serializer = HotelSerializer(
+            available_hotels, many=True, context={"request": request}
+        )
+        return Response(serializer.data, status=status.HTTP_200_OK)
+
+    @action(
+        detail=True,
+        methods=["get"],
+        url_path="available-room-types",
```
Comment (rafcasceb) [inline backend/pawtel/hotels/views.py:132]: Aquí que sea el url path:  `room-types/available`.

**Review by rafcasceb — CHANGES_REQUESTED** [review]

**Review by rafcasceb — COMMENTED** [review]
Arreglar error de conventional commits y el warning de las 51 líneas.
